src/WorkflowCore/Services/WorkflowMiddlewareRunner.cs

```diff
@@ -0,0 +1,106 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using WorkflowCore.Interface;
+using WorkflowCore.Models;
+
+namespace WorkflowCore.Services
+{
+    /// <summary>
+    /// Runner with the ability to apply middleware to steps and workflows.
+    /// </summary>
+    public class WorkflowMiddlewareRunner : IWorkflowMiddlewareRunner
```

Comment:
As it stands now the workflow middleware runner does execute the step as the last "`next`" in the step middleware chain. This is deliberate to allow step middleware to add logic around the execution of a step and even potentially change the step's result.

Are you suggesting to not have it run the step?

Reply:
No, I think we it might help if the name of the class indicated that it also ran the step?

Reply:
Ah gotcha. Sounds good. Will rename to reflect that. Something like `WorkflowMiddlewareAndStepRunner`? Or perhaps `WorkflowStepRunner` and the middleware portion is implied?

Reply:
Hmm... Thinking about this some more. `WorkflowMiddlewareAndStepRunner` doesn't exactly sound well. Then I tried changing it to `WorkflowStepRunner` but that doesn't imply that it also runs pre/post workflow middleware. So regarding naming, I think we have these options:

1. Keep it as `WorkflowMiddlewareRunner` and document that it also runs the step in the method name and comments. We could always rename `RunStep` -> `RunStepWithMiddleware` to emphasize that it runs the step and also runs middleware around it.
2. Split the class into two classes, one for running pre/post middleware and the other for running steps with middleware. I'd probably call them `WorkflowMiddlewareRunner` and `WorkflowStepRunner` respectively. Although the major con here is that there will be two dependencies needed to be added to the executor constructor instead of one.

I'm in favor of option 1. What are your thoughts?

Reply:
I feel like the main objective is to run the step, and running the middleware is a secondary part of that, so in that case I would opt to name it something like `StepExecutor` and keep the functionality as you have it?

Reply:
Sounds good. I like `StepExecutor`

Reply:
With this being called `StepExecutor` now, I think it makes even more sense to split it into two classes, one for steps and one for workflow pre/post middleware. Do you agree?

Reply:
Take a look at my latest changes, I split it out and it made testing much easier and follows the single responsibility principle quite nicely. Now there's `StepExecutor` which executes the step with middleware and `WorkflowMiddlewareRunner` ONLY runs pre/post workflow. This felt the most natural.